easybuild/hooks/reframe_hook.py

```diff
@@ -0,0 +1,262 @@
+    def prepend_fake_module_path():
+        """modified from easybuild.framework.easyblock.Easyblock.load_fake_module function"""
+        env = copy.deepcopy(os.environ)
+        fake_mod_path = self.make_module_step(fake=True)
+        self.modules_tool.prepend_module_path(os.path.join(fake_mod_path, self.mod_subdir), priority=10000)
+        return (fake_mod_path, env)
+
+    if self.name == 'reframe':
+        # Run ReFrame unittests?
+        return
+    else:
+        # Load reframe module
+        self.modules_tool.load(["reframe"], allow_reload=False)
+
+    # Probe relevant informations to select reframe tests
+    CUDA = get_software_root("CUDA")
+
+    # Prepend newly created module to MODULEPATH
+    dry_run = build_option('extended_dry_run')
+    silent = build_option('silent')
+    if not dry_run:
+        fake_mod_data = prepend_fake_module_path()
+
+    random_int = random.randint(0, 9999999)
+    random_folder = "tmp-%07d" % random_int
+
+    # Set ReFrame output and perflogs folders
+    output_dir = os.path.join(self.installdir, log_path(), 'reframe_output')
+    _log.info("ReFrame output dir: %s", output_dir)
+
+    perflogs_dir = os.path.join(self.installdir, log_path(), 'reframe_perflogs')
+    _log.info("ReFrame perflogs dir: %s", perflogs_dir)
+
+    # Set ReFrame run and stage dir
+    # stage dir cannot be self.builddir because on daint it is /run/user
+    # which is a non shared filesystem with the compute nodes
+    stage_dir = os.path.join(self.builddir, log_path(), 'reframe_stage')
+    rfm_run_dir = self.builddir
+    scratch = os.getenv('SCRATCH', None)
+    if scratch:
+        stage_dir = os.path.join(scratch, log_path(), random_folder, 'reframe_stage')
+        rfm_run_dir = os.path.join(scratch, log_path(), random_folder, 'rfm_run_dir')
+    elif self.build_in_installdir and not dry_run:
+        # generating self.builddir
+        self.gen_builddir()
+        rfm_run_dir = self.builddir
+        stage_dir = os.path.join(self.builddir, log_path(), random_folder, 'reframe_stage')
+        # restoring the self.builddir status
+        self.builddir = self.installdir
+
+    _log.info("ReFrame stage dir: %s", stage_dir)
+    _log.info("ReFrame run dir: %s", rfm_run_dir)
+
+    # creating run dir if it does not exist
+    if not os.path.exists(rfm_run_dir):
+        os.makedirs(rfm_run_dir)
+
+    if dry_run:
+        dry_run_msg("ReFrame dirs:", silent=silent)
+        dry_run_msg("  * run: %s" % rfm_run_dir, silent=silent)
+        dry_run_msg("  * output: %s" % output_dir, silent=silent)
+        dry_run_msg("  * perflogs: %s" % perflogs_dir, silent=silent)
+        dry_run_msg("  * stage: %s" % stage_dir, silent=silent)
+
+    # add to rfm_cmd your common system flags
+    rfm_cmd = "reframe --nocolor -r "
+    rfm_cmd += "-o %s " % output_dir
+    rfm_cmd += "-s %s " % stage_dir
+    rfm_cmd += "--perflogdir %s " % perflogs_dir
+    rfm_cmd += "--save-log-files "
+
+    if self.name == 'GROMACS' and CUDA:
```

Comment:
I understand that this is just a POC, but are you planning to have a more structured/maintainable way of storing the EB vs. ReFrame mappings?

Reply:
Yes. This is just a POC. We plan to have another file which will have a dictionary that will map the system with the tests.

```
'reframe': {
    'GROMACSGPU' : {
        'descr': 'GROMACS CUDA tests',
        'name': ['GROMACS'], # if easyconfig name is GROMACS
        'dependency': ['CUDA'], # translated as get_software_root("CUDA")
        'tests': ["gromacs_gpu_prod_check"], # list of ReFrame tests 
        'module_to_replace': ['GROMACS'] # name of the module inside the test, if none, name will be taken from the 'name' key
    },
    'GROMACSCPU' : {
        'descr': 'GROMACS CPU only tests',
        'name': ['GROMACS'], # if easyconfig name is GROMACS
        'dependency': [], # translated as nothing
        'tests': ["gromacs_cpu_prod_check"], # list of ReFrame tests 
        'module_to_replace': ['GROMACS'] # name of the module inside the test, if none, name will be taken from the 'name' key
    },
    'PrgEnvs' : {
        'descr': 'Basic PrgEnv tests',
        'name': ['gmpolf', 'foss', 'gmvolf', 'CrayCCE', 'CrayGNU', 'CrayIntel' 'CrayPGI'], # if easyconfig name is one of the ones inside
        'dependency': [], 
        'tests': ["HelloWorldTestSerial_cpp_dynamic", "HelloWorldTestSerial_cpp_static", "HelloWorldTestSerial_c_dynamic", "HelloWorldTestSerial_c_static", "HelloWorldTestSerial_f90_dynamic", "HelloWorldTestSerial_f90_static", "HelloWorldTestOpenMP_cpp_dynamic", "HelloWorldTestOpenMP_cpp_static", "HelloWorldTestOpenMP_c_dynamic", "HelloWorldTestOpenMP_c_static", "HelloWorldTestOpenMP_f90_dynamic", "HelloWorldTestOpenMP_f90_static", "HelloWorldTestMPI_cpp_dynamic", "HelloWorldTestMPI_cpp_static", "HelloWorldTestMPI_c_dynamic", "HelloWorldTestMPI_c_static", "HelloWorldTestMPI_f90_dynamic", "HelloWorldTestMPI_f90_static", "HelloWorldTestMPIOpenMP_cpp_dynamic", "HelloWorldTestMPIOpenMP_cpp_static", "HelloWorldTestMPIOpenMP_c_dynamic", "HelloWorldTestMPIOpenMP_c_static", "HelloWorldTestMPIOpenMP_f90_dynamic", "HelloWorldTestMPIOpenMP_f90_static", "jacobi_C", "jacobi_Cpp", "jacobi_F9", "DGEMMTest", "ScaLAPACKSanity_static", "ScaLAPACKSanity_dynamic", "mpi_helloworld_check"], # list of ReFrame tests 
        'module_to_replace': [] # name of the module inside the test, if none, name will be taken from the 'name' key
    },
},
```